Upload File

In [ ]:
from google.colab import files

uploaded = files.upload()

Import Libraries

In [ ]:
import pandas as pd
import numpy as np
from math import sqrt

Reading Data

In [ ]:
orders = pd.read_csv("RH Customer Orders  Dataset_8.csv")
products = pd.read_csv("Products.csv")

print("Orders shape:", orders.shape)
print("Products shape:", products.shape)

display(orders.head())
display(products.head())

Orders shape: (194, 4)
Products shape: (43, 9)


,Customer ID,Brand,Product Description,Quantity
0,CUST5271,Aero Wood Round Dining Table,50'' Bistro Table,1
1,CUST5271,Nicola Track Vegan Leather Dining Side Chair,Dining Chair,4
2,CUST5271,Byron Canopy Bed,Cal King Bed 102'',1
3,CUST5271,Noemi Torchiere Floor Lamp,Burnished Brass,1
4,CUST5271,Maxwell Leather Right-Arm Bench-Seat L-Sectional,Luxe 108'' Sectional,1


,Item,Retail Price,Variant,CubicFeet,WeightLbs,URL,Item Type,Loading Time (mins),Unloading & Staging Time (mins)
0,Byron Emperador Dining Table,5330,72'' Table,50.00,462,https://rh.com/us/en/catalog/product/product.j...,Dining & Fire Table,3.0,5.0
1,Byron Emperador Dining Table,5330,84'' Table,58.33,512,https://rh.com/us/en/catalog/product/product.j...,Dining & Fire Table,3.0,5.0
2,Byron Emperador Dining Table,5330,96'' Table,66.67,484,https://rh.com/us/en/catalog/product/product.j...,Dining & Fire Table,3.0,5.0
3,Aero Wood Round Dining Table,1860,30'' Bistro Table,12.27,31,https://rh.com/us/en/catalog/product/product.j...,Dining & Fire Table,3.0,5.0
4,Aero Wood Round Dining Table,1860,36'' Bistro Table,17.67,51,https://rh.com/us/en/catalog/product/product.j...,Dining & Fire Table,3.0,5.0


look at the column names

In [ ]:
print("Orders columns:")
print(orders.columns.tolist())

print("\nProducts columns:")
print(products.columns.tolist())

Orders columns:
['Customer ID', 'Brand', 'Product Description', 'Quantity']

Products columns:
['Item', 'Retail Price', 'Variant', 'CubicFeet', 'WeightLbs', 'URL', 'Item Type', 'Loading Time (mins)', 'Unloading & Staging Time (mins)']


Clean Up Fields

In [ ]:
orders.columns = orders.columns.str.strip()
products.columns = products.columns.str.strip()

orders["Product Description"] = orders["Product Description"].astype(str).str.strip()
products["Item"] = products["Item"].astype(str).str.strip()
products["Variant"] = products["Variant"].astype(str).str.strip()
products["Item Type"] = products["Item Type"].astype(str).str.strip()

List Merging (Please do not run this for now.)

In [ ]:
orders.merge(products, left_on="Product Description", right_on="Item")

,Customer ID,Brand,Product Description,Quantity,Item,Retail Price,Variant,CubicFeet,WeightLbs,URL,Item Type,Loading Time (mins),Unloading & Staging Time (mins)


Create a Mapping Table

In [ ]:
mapping = {
    "72'' Dining Table": ("Byron Emperador Dining Table", "72'' Table"),
    "84'' Dining Table": ("Byron Emperador Dining Table", "84'' Table"),
    "96'' Dining Table": ("Byron Emperador Dining Table", "96'' Table"),

    "30'' Bistro Table": ("Aero Wood Round Dining Table", "30'' Bistro Table"),
    "36'' Bistro Table": ("Aero Wood Round Dining Table", "36'' Bistro Table"),
    "42'' Bistro Table": ("Aero Wood Round Dining Table", "42'' bistro Table"),
    "50'' Bistro Table": ("Aero Wood Round Dining Table", "50'' Bistro Table"),

    "48'' Table": ("Reclaimed Oak Reeded Round Dining Table", "48'' Table"),
    "60'' Table": ("Reclaimed Oak Reeded Round Dining Table", "60'' Table"),
    "72'' Table": ("Reclaimed Oak Reeded Round Dining Table", "72'' Table"),

    "Dining Armchair": ("Arno Fabric Dining Armchair", "Standard"),
    "Dining Side Chair": ("Nicola Track Vegan Leather Dining Side Chair", "Standard"),
    "Dining Chair": ("Arno Vegan Leather Dining Side Chair", "Standard"),

    "Barstool 40 1/4'' H": ("Arno Fabric Bar & Counter Stool", "Barstool 40 1/4'' H"),
    "Counter Stool 36 1/4'' H": ("Arno Fabric Bar & Counter Stool", "Counter Stool 36 1/4'' H"),

    "Bar Cabinet": ("Melbourne Bar Cabinet", "Standard"),

    "Classic 6' Sofa": ("Maxwell Leather Sofa", "Classic 6'"),
    "Classic 8' Sofa": ("Maxwell Leather Sofa", "Classic 8'"),
    "Luxe 7' Sofa": ("Maxwell Leather Sofa", "Luxe 7'"),
    "Luxe 9' Sofa": ("Maxwell Leather Sofa", "Luxe 9'"),

    "Classic 72'' Sectional": ("Maxwell Leather Right-Arm Bench-Seat L-Sectional", "Classic 72''"),
    "Classic 96'' Sectional": ("Maxwell Leather Right-Arm Bench-Seat L-Sectional", "Classic 96''"),
    "Luxe 108'' Sectional": ("Maxwell Leather Right-Arm Bench-Seat L-Sectional", "Luxe 108''"),
    "Luxe 12'' Sectional": ("Maxwell Leather Right-Arm Bench-Seat L-Sectional", "Luxe 12''"),

    "King Bed 90''": ("Byron Canopy Bed", "King Bed 90''"),
    "Cal King Bed 90''": ("Byron Canopy Bed", "Cal King Bed 90 ''"),
    "King Bed 102''": ("Byron Canopy Bed", "King Bed 102''"),
    "Cal King Bed 102''": ("Byron Canopy Bed", "Cal King Bed 102''"),

    "Queen Bed": ("Ligné Panel Bed with Footboard", "Queen Bed"),
    "King Bed": ("Ligné Panel Bed with Footboard", "King Bed"),
    "Cal King Bed": ("Ligné Panel Bed with Footboard", "Cal King Bed"),

    "18'' Nightstand": ("Kyoto Closed Nightstand", "18'' nightstand"),
    "22'' Nightstand": ("Kyoto Closed Nightstand", "22'' nightstand"),
    "26'' Nightstand": ("Kyoto Closed Nightstand", "26'' nightstand"),
    "32'' Nightstand": ("Kyoto Closed Nightstand", "32'' nightstand"),

    "19'' Lamp": ("Mycena Obelisk Stone Table Lamp", "19'' Lamp"),
    "24'' Lamp": ("Mycena Obelisk Stone Table Lamp", "24'' Lamp"),

    "Burnished Brass": ("Noemi Torchiere Floor Lamp", "Burnished Brass"),
    "Carbon Fountain": ('Terreno Wall Fountain 22"', "Carbon Fountain"),

    "56'' W Fire Table": ("Orizaba Rectangular Fire Table", "56'' W fire table"),
    "72'' W Fire Table": ("Orizaba Rectangular Fire Table", "72'' W fire table"),
}

Apply the mapping to the orders.

In [ ]:
orders["Mapped_Item"] = orders["Product Description"].map(lambda x: mapping.get(x, (None, None))[0])
orders["Mapped_Variant"] = orders["Product Description"].map(lambda x: mapping.get(x, (None, None))[1])

display(orders.head(10))

,Customer ID,Brand,Product Description,Quantity,Mapped_Item,Mapped_Variant
0,CUST5271,Aero Wood Round Dining Table,50'' Bistro Table,1,Aero Wood Round Dining Table,50'' Bistro Table
1,CUST5271,Nicola Track Vegan Leather Dining Side Chair,Dining Chair,4,Arno Vegan Leather Dining Side Chair,Standard
2,CUST5271,Byron Canopy Bed,Cal King Bed 102'',1,Byron Canopy Bed,Cal King Bed 102''
3,CUST5271,Noemi Torchiere Floor Lamp,Burnished Brass,1,Noemi Torchiere Floor Lamp,Burnished Brass
4,CUST5271,Maxwell Leather Right-Arm Bench-Seat L-Sectional,Luxe 108'' Sectional,1,Maxwell Leather Right-Arm Bench-Seat L-Sectional,Luxe 108''
5,CUST5271,Kyoto Closed Nightstand,26'' Nightstand,2,Kyoto Closed Nightstand,26'' nightstand
6,CUST5271,Maxwell Leather Sofa,Luxe 7' Sofa,1,Maxwell Leather Sofa,Luxe 7'
7,CUST5981,Reclaimed Oak Reeded Round Dining Table,72'' Table,1,Reclaimed Oak Reeded Round Dining Table,72'' Table
8,CUST5981,Nicola Track Vegan Leather Dining Side Chair,Dining Chair,8,Arno Vegan Leather Dining Side Chair,Standard
9,CUST5981,Orizaba Rectangular Fire Table,56'' W Fire Table,2,Orizaba Rectangular Fire Table,56'' W fire table


###Check for any mismatches.
###If the output is empty, it indicates that the mapping was successful.

In [ ]:
unmatched = orders[orders["Mapped_Item"].isna()]["Product Description"].unique()
print("Unmatched values:")
print(unmatched)

Unmatched values:
[]


Merge the Orders and Products tables.

In [ ]:
df = orders.merge(
    products,
    left_on=["Mapped_Item", "Mapped_Variant"],
    right_on=["Item", "Variant"],
    how="left"
)

print("Merged shape:", df.shape)
display(df.head())

Merged shape: (194, 15)


,Customer ID,Brand,Product Description,Quantity,Mapped_Item,Mapped_Variant,Item,Retail Price,Variant,CubicFeet,WeightLbs,URL,Item Type,Loading Time (mins),Unloading & Staging Time (mins)
0,CUST5271,Aero Wood Round Dining Table,50'' Bistro Table,1,Aero Wood Round Dining Table,50'' Bistro Table,Aero Wood Round Dining Table,1860,50'' Bistro Table,33.52,72,https://rh.com/us/en/catalog/product/product.j...,Dining & Fire Table,3.0,5.0
1,CUST5271,Nicola Track Vegan Leather Dining Side Chair,Dining Chair,4,Arno Vegan Leather Dining Side Chair,Standard,Arno Vegan Leather Dining Side Chair,395,Standard,7.87,15,https://rh.com/us/en/catalog/product/product.j...,Chair,1.0,2.0
2,CUST5271,Byron Canopy Bed,Cal King Bed 102'',1,Byron Canopy Bed,Cal King Bed 102'',Byron Canopy Bed,9330,Cal King Bed 102'',414.38,507,https://rh.com/us/en/catalog/product/product.j...,Bed Frame,6.0,8.0
3,CUST5271,Noemi Torchiere Floor Lamp,Burnished Brass,1,Noemi Torchiere Floor Lamp,Burnished Brass,Noemi Torchiere Floor Lamp,3330,Burnished Brass,7.18,21,https://rh.com/us/en/catalog/product/product.j...,Lamp,1.5,2.0
4,CUST5271,Maxwell Leather Right-Arm Bench-Seat L-Sectional,Luxe 108'' Sectional,1,Maxwell Leather Right-Arm Bench-Seat L-Sectional,Luxe 108'',Maxwell Leather Right-Arm Bench-Seat L-Sectional,13990,Luxe 108'',97.75,684,https://rh.com/us/en/catalog/product/product.j...,Sofa,2.5,4.0


Check for any failed merges.

In [ ]:
missing_rows = df[df["CubicFeet"].isna()]
print("Missing merged rows:", len(missing_rows))
display(missing_rows[["Customer ID", "Product Description", "Mapped_Item", "Mapped_Variant"]].drop_duplicates())

Missing merged rows: 0


,Customer ID,Product Description,Mapped_Item,Mapped_Variant


Calculate the volume, weight, and time for each order line.

In [ ]:
df["Total Volume"] = df["CubicFeet"] * df["Quantity"]
df["Total Weight"] = df["WeightLbs"] * df["Quantity"]
df["Total Load Time"] = df["Loading Time (mins)"] * df["Quantity"]
df["Total Unload Time"] = df["Unloading & Staging Time (mins)"] * df["Quantity"]

display(df[[
    "Customer ID", "Product Description", "Quantity",
    "CubicFeet", "WeightLbs", "Total Volume", "Total Weight"
]].head())

,Customer ID,Product Description,Quantity,CubicFeet,WeightLbs,Total Volume,Total Weight
0,CUST5271,50'' Bistro Table,1,33.52,72,33.52,72
1,CUST5271,Dining Chair,4,7.87,15,31.48,60
2,CUST5271,Cal King Bed 102'',1,414.38,507,414.38,507
3,CUST5271,Burnished Brass,1,7.18,21,7.18,21
4,CUST5271,Luxe 108'' Sectional,1,97.75,684,97.75,684


Create a Client-Level Summary Table

In [ ]:
customer_orders = df.groupby("Customer ID").agg(
    Total_Volume=("Total Volume", "sum"),
    Total_Weight=("Total Weight", "sum"),
    Total_Load_Time=("Total Load Time", "sum"),
    Total_Unload_Time=("Total Unload Time", "sum"),
    Total_Items=("Quantity", "sum")
).reset_index()

customer_orders["Total_Service_Time"] = customer_orders["Total_Load_Time"] + customer_orders["Total_Unload_Time"]

print("Number of customers:", customer_orders.shape[0])
display(customer_orders.head(10))

Number of customers: 19


,Customer ID,Total_Volume,Total_Weight,Total_Load_Time,Total_Unload_Time,Total_Items,Total_Service_Time
0,CUST1667,1024.68,5053,56.0,100.0,38,156.0
1,CUST2208,189.05,464,9.0,13.0,3,22.0
2,CUST2289,1274.12,4211,59.5,96.0,33,155.5
3,CUST3051,1860.07,5404,81.5,147.0,43,228.5
4,CUST3446,2133.54,5904,83.5,155.5,35,239.0
5,CUST4225,343.33,1815,27.5,51.0,20,78.5
6,CUST4340,980.88,5112,65.0,110.5,37,175.5
7,CUST4472,16.33,325,3.0,5.0,1,8.0
8,CUST5271,684.20,2056,22.5,36.0,11,58.5
9,CUST5533,876.78,3685,48.0,79.5,24,127.5


#Check for any customers whose orders exceed the truck's capacity.

In [ ]:
MAX_VOLUME = 1700
MAX_WEIGHT = 7000

oversized_customers = customer_orders[
    (customer_orders["Total_Volume"] > MAX_VOLUME) |
    (customer_orders["Total_Weight"] > MAX_WEIGHT)
]

print("Oversized customers:", len(oversized_customers))
display(oversized_customers)

Oversized customers: 5


,Customer ID,Total_Volume,Total_Weight,Total_Load_Time,Total_Unload_Time,Total_Items,Total_Service_Time
3,CUST3051,1860.07,5404,81.5,147.0,43,228.5
4,CUST3446,2133.54,5904,83.5,155.5,35,239.0
11,CUST5981,2414.60,6439,80.0,129.5,35,209.5
12,CUST6434,1898.57,5742,65.0,103.0,28,168.0
15,CUST8286,1726.42,6721,62.0,98.5,30,160.5


#Read Customer Address Table

In [ ]:
addresses = pd.read_excel("Unlimited Furniture Delivery at RH - Customers Addresses.xlsx")

print("Addresses shape:", addresses.shape)
print(addresses.columns.tolist())

display(addresses.head(10))

Addresses shape: (372, 2)
['Customer ID', 'Customer Address']


,Customer ID,Customer Address
0,CUST1011,"457 Diller Street, Alameda, CA 94501"
1,CUST1064,"2164 San Jose Avenue, Alameda, CA 94501"
2,CUST1070,"1535 Saint Charles Street, Alameda, CA 94501"
3,CUST1075,"85 Vicente Rd, Berkeley, CA 94705"
4,CUST1133,13397 Campus Drive – 94619 Oakland
5,CUST1155,"945 Green St UNIT 8, San Francisco, CA 94133"
6,CUST1162,"1325 Glendale Ave, Berkeley, CA 94708"
7,CUST1174,"105 3rd Street #2, Sausalito, CA 94965"
8,CUST1201,"101 Sacramento Way, Sausalito, CA 94965"
9,CUST1202,"765 Eagle Avenue, Alameda, CA 94501"


#Clean up the column names.



In [ ]:
addresses.columns = addresses.columns.str.strip()

display(addresses.head(10))

,Customer ID,Customer Address
0,CUST1011,"457 Diller Street, Alameda, CA 94501"
1,CUST1064,"2164 San Jose Avenue, Alameda, CA 94501"
2,CUST1070,"1535 Saint Charles Street, Alameda, CA 94501"
3,CUST1075,"85 Vicente Rd, Berkeley, CA 94705"
4,CUST1133,13397 Campus Drive – 94619 Oakland
5,CUST1155,"945 Green St UNIT 8, San Francisco, CA 94133"
6,CUST1162,"1325 Glendale Ave, Berkeley, CA 94708"
7,CUST1174,"105 3rd Street #2, Sausalito, CA 94965"
8,CUST1201,"101 Sacramento Way, Sausalito, CA 94965"
9,CUST1202,"765 Eagle Avenue, Alameda, CA 94501"


#Standardize the Customer ID format.

In [ ]:
customer_orders["Customer ID"] = customer_orders["Customer ID"].astype(str).str.strip()
addresses["Customer ID"] = addresses["Customer ID"].astype(str).str.strip()

#merge

In [ ]:
customer_orders = customer_orders.merge(
    addresses,
    on="Customer ID",
    how="left"
)

print("After merge:", customer_orders.shape)
display(customer_orders.head(10))

After merge: (19, 8)


,Customer ID,Total_Volume,Total_Weight,Total_Load_Time,Total_Unload_Time,Total_Items,Total_Service_Time,Customer Address
0,CUST1667,1024.68,5053,56.0,100.0,38,156.0,"Garden Plan, San Leandro, CA 94578"
1,CUST2208,189.05,464,9.0,13.0,3,22.0,"912 Sienna Loop, San Leandro, CA 94579"
2,CUST2289,1274.12,4211,59.5,96.0,33,155.5,"4 Captain Dr Unit E211, Emeryville, CA 94608"
3,CUST3051,1860.07,5404,81.5,147.0,43,228.5,"1510 165th Ave, San Leandro, CA 94578"
4,CUST3446,2133.54,5904,83.5,155.5,35,239.0,"1515 165th Ave, San Leandro, CA 94578"
5,CUST4225,343.33,1815,27.5,51.0,20,78.5,"2125 166th Ave, San Leandro, CA 94578"
6,CUST4340,980.88,5112,65.0,110.5,37,175.5,"28 Buckelew Street, Sausalito, CA 94965"
7,CUST4472,16.33,325,3.0,5.0,1,8.0,"1530 165th Ave, San Leandro, CA 94578"
8,CUST5271,684.20,2056,22.5,36.0,11,58.5,"8234 Jackson St, San Francisco, CA 94109 (Nob..."
9,CUST5533,876.78,3685,48.0,79.5,24,127.5,"1415 La Loma Ave, Berkeley, CA 94708"


#Check for any unmatched clients.

In [ ]:
missing_address = customer_orders[customer_orders.isna().any(axis=1)]
print("Rows with some missing fields:", len(missing_address))

display(missing_address.head(20))

Rows with some missing fields: 0


,Customer ID,Total_Volume,Total_Weight,Total_Load_Time,Total_Unload_Time,Total_Items,Total_Service_Time,Customer Address


#Using Address Tables

In [ ]:
addresses = pd.read_excel("Unlimited Furniture Delivery at RH - Customers Addresses.xlsx")
addresses.columns = addresses.columns.str.strip()

print(addresses.shape)
print(addresses.columns.tolist())
display(addresses.head(10))

(372, 2)
['Customer ID', 'Customer Address']


,Customer ID,Customer Address
0,CUST1011,"457 Diller Street, Alameda, CA 94501"
1,CUST1064,"2164 San Jose Avenue, Alameda, CA 94501"
2,CUST1070,"1535 Saint Charles Street, Alameda, CA 94501"
3,CUST1075,"85 Vicente Rd, Berkeley, CA 94705"
4,CUST1133,13397 Campus Drive – 94619 Oakland
5,CUST1155,"945 Green St UNIT 8, San Francisco, CA 94133"
6,CUST1162,"1325 Glendale Ave, Berkeley, CA 94708"
7,CUST1174,"105 3rd Street #2, Sausalito, CA 94965"
8,CUST1201,"101 Sacramento Way, Sausalito, CA 94965"
9,CUST1202,"765 Eagle Avenue, Alameda, CA 94501"


In [ ]:
customer_orders["Customer ID"] = customer_orders["Customer ID"].astype(str).str.strip()
addresses["Customer ID"] = addresses["Customer ID"].astype(str).str.strip()

customer_orders = customer_orders.merge(addresses, on="Customer ID", how="left")

print(customer_orders.shape)
display(customer_orders.head(10))

(19, 9)


,Customer ID,Total_Volume,Total_Weight,Total_Load_Time,Total_Unload_Time,Total_Items,Total_Service_Time,Customer Address_x,Customer Address_y
0,CUST1667,1024.68,5053,56.0,100.0,38,156.0,"Garden Plan, San Leandro, CA 94578","Garden Plan, San Leandro, CA 94578"
1,CUST2208,189.05,464,9.0,13.0,3,22.0,"912 Sienna Loop, San Leandro, CA 94579","912 Sienna Loop, San Leandro, CA 94579"
2,CUST2289,1274.12,4211,59.5,96.0,33,155.5,"4 Captain Dr Unit E211, Emeryville, CA 94608","4 Captain Dr Unit E211, Emeryville, CA 94608"
3,CUST3051,1860.07,5404,81.5,147.0,43,228.5,"1510 165th Ave, San Leandro, CA 94578","1510 165th Ave, San Leandro, CA 94578"
4,CUST3446,2133.54,5904,83.5,155.5,35,239.0,"1515 165th Ave, San Leandro, CA 94578","1515 165th Ave, San Leandro, CA 94578"
5,CUST4225,343.33,1815,27.5,51.0,20,78.5,"2125 166th Ave, San Leandro, CA 94578","2125 166th Ave, San Leandro, CA 94578"
6,CUST4340,980.88,5112,65.0,110.5,37,175.5,"28 Buckelew Street, Sausalito, CA 94965","28 Buckelew Street, Sausalito, CA 94965"
7,CUST4472,16.33,325,3.0,5.0,1,8.0,"1530 165th Ave, San Leandro, CA 94578","1530 165th Ave, San Leandro, CA 94578"
8,CUST5271,684.20,2056,22.5,36.0,11,58.5,"8234 Jackson St, San Francisco, CA 94109 (Nob...","8234 Jackson St, San Francisco, CA 94109 (Nob..."
9,CUST5533,876.78,3685,48.0,79.5,24,127.5,"1415 La Loma Ave, Berkeley, CA 94708","1415 La Loma Ave, Berkeley, CA 94708"


#Install geopy

In [ ]:
!pip install geopy

#Convert an address to latitude and longitude.

In [ ]:
from geopy.geocoders import Nominatim
from geopy.extra.rate_limiter import RateLimiter

geolocator = Nominatim(user_agent="rh_delivery_project")
geocode = RateLimiter(geolocator.geocode, min_delay_seconds=1)

#Check existing column names.

In [ ]:
print(customer_orders.columns.tolist())

['Customer ID', 'Total_Volume', 'Total_Weight', 'Total_Load_Time', 'Total_Unload_Time', 'Total_Items', 'Total_Service_Time', 'Customer Address_x', 'Customer Address_y']


#Standardized Column Names

In [ ]:
customer_orders = customer_orders.rename(columns={"Customer Address_x": "Customer Address"})

#Perform geocoding

In [ ]:
customer_orders["Location"] = customer_orders["Customer Address"].apply(geocode)

customer_orders["Lat"] = customer_orders["Location"].apply(
    lambda loc: loc.latitude if loc else None
)
customer_orders["Lon"] = customer_orders["Location"].apply(
    lambda loc: loc.longitude if loc else None
)

display(customer_orders[[
    "Customer ID", "Customer Address", "Lat", "Lon"
]])

,Customer ID,Customer Address,Lat,Lon
0,CUST1667,"Garden Plan, San Leandro, CA 94578",NaN,NaN
1,CUST2208,"912 Sienna Loop, San Leandro, CA 94579",37.692196,-122.146782
2,CUST2289,"4 Captain Dr Unit E211, Emeryville, CA 94608",NaN,NaN
3,CUST3051,"1510 165th Ave, San Leandro, CA 94578",37.694998,-122.109606
4,CUST3446,"1515 165th Ave, San Leandro, CA 94578",NaN,NaN
5,CUST4225,"2125 166th Ave, San Leandro, CA 94578",37.698877,-122.102436
6,CUST4340,"28 Buckelew Street, Sausalito, CA 94965",NaN,NaN
7,CUST4472,"1530 165th Ave, San Leandro, CA 94578",37.695048,-122.109281
8,CUST5271,"8234 Jackson St, San Francisco, CA 94109 (Nob...",37.794357,-122.419773
9,CUST5533,"1415 La Loma Ave, Berkeley, CA 94708",37.883312,-122.258016


#As can be seen, several coordinates are missing from the data above; therefore, we will first identify and isolate these missing coordinates.

In [ ]:
missing_geo = customer_orders[customer_orders["Lat"].isna()]
print("Missing geocoding:", len(missing_geo))
display(missing_geo[["Customer ID", "Customer Address"]])

Missing geocoding: 5


,Customer ID,Customer Address
0,CUST1667,"Garden Plan, San Leandro, CA 94578"
2,CUST2289,"4 Captain Dr Unit E211, Emeryville, CA 94608"
4,CUST3446,"1515 165th Ave, San Leandro, CA 94578"
6,CUST4340,"28 Buckelew Street, Sausalito, CA 94965"
12,CUST6434,"1 Captain Dr Unit D255, Emeryville, CA 94608"


#Create a clean version of the address.

In [ ]:
import re

def clean_address(addr):
    if pd.isna(addr):
        return addr

    addr = str(addr)

    # Remove the content within the parentheses.
    addr = re.sub(r"\(.*?\)", "", addr)

    # Remove the content following "Unit / Apt / Suite".
    addr = re.sub(r"\b(Unit|Apt|Apartment|Suite)\b.*?,", ",", addr, flags=re.IGNORECASE)

    # Remove extra spaces.
    addr = re.sub(r"\s+", " ", addr).strip()

    return addr

customer_orders["Clean_Address"] = customer_orders["Customer Address"].apply(clean_address)

display(customer_orders[["Customer ID", "Customer Address", "Clean_Address"]])

,Customer ID,Customer Address,Clean_Address
0,CUST1667,"Garden Plan, San Leandro, CA 94578","Garden Plan, San Leandro, CA 94578"
1,CUST2208,"912 Sienna Loop, San Leandro, CA 94579","912 Sienna Loop, San Leandro, CA 94579"
2,CUST2289,"4 Captain Dr Unit E211, Emeryville, CA 94608","4 Captain Dr , Emeryville, CA 94608"
3,CUST3051,"1510 165th Ave, San Leandro, CA 94578","1510 165th Ave, San Leandro, CA 94578"
4,CUST3446,"1515 165th Ave, San Leandro, CA 94578","1515 165th Ave, San Leandro, CA 94578"
5,CUST4225,"2125 166th Ave, San Leandro, CA 94578","2125 166th Ave, San Leandro, CA 94578"
6,CUST4340,"28 Buckelew Street, Sausalito, CA 94965","28 Buckelew Street, Sausalito, CA 94965"
7,CUST4472,"1530 165th Ave, San Leandro, CA 94578","1530 165th Ave, San Leandro, CA 94578"
8,CUST5271,"8234 Jackson St, San Francisco, CA 94109 (Nob...","8234 Jackson St, San Francisco, CA 94109"
9,CUST5533,"1415 La Loma Ave, Berkeley, CA 94708","1415 La Loma Ave, Berkeley, CA 94708"


#Re-geocode only the failed items.

In [ ]:
mask_missing = customer_orders["Lat"].isna()

customer_orders.loc[mask_missing, "Location_retry"] = customer_orders.loc[mask_missing, "Clean_Address"].apply(geocode)

customer_orders.loc[mask_missing, "Lat"] = customer_orders.loc[mask_missing, "Location_retry"].apply(
    lambda loc: loc.latitude if loc else None
)

customer_orders.loc[mask_missing, "Lon"] = customer_orders.loc[mask_missing, "Location_retry"].apply(
    lambda loc: loc.longitude if loc else None
)

display(customer_orders[["Customer ID", "Customer Address", "Clean_Address", "Lat", "Lon"]])

,Customer ID,Customer Address,Clean_Address,Lat,Lon
0,CUST1667,"Garden Plan, San Leandro, CA 94578","Garden Plan, San Leandro, CA 94578",NaN,NaN
1,CUST2208,"912 Sienna Loop, San Leandro, CA 94579","912 Sienna Loop, San Leandro, CA 94579",37.692196,-122.146782
2,CUST2289,"4 Captain Dr Unit E211, Emeryville, CA 94608","4 Captain Dr , Emeryville, CA 94608",37.838388,-122.303434
3,CUST3051,"1510 165th Ave, San Leandro, CA 94578","1510 165th Ave, San Leandro, CA 94578",37.694998,-122.109606
4,CUST3446,"1515 165th Ave, San Leandro, CA 94578","1515 165th Ave, San Leandro, CA 94578",NaN,NaN
5,CUST4225,"2125 166th Ave, San Leandro, CA 94578","2125 166th Ave, San Leandro, CA 94578",37.698877,-122.102436
6,CUST4340,"28 Buckelew Street, Sausalito, CA 94965","28 Buckelew Street, Sausalito, CA 94965",NaN,NaN
7,CUST4472,"1530 165th Ave, San Leandro, CA 94578","1530 165th Ave, San Leandro, CA 94578",37.695048,-122.109281
8,CUST5271,"8234 Jackson St, San Francisco, CA 94109 (Nob...","8234 Jackson St, San Francisco, CA 94109",37.794357,-122.419773
9,CUST5533,"1415 La Loma Ave, Berkeley, CA 94708","1415 La Loma Ave, Berkeley, CA 94708",37.883312,-122.258016


#Re-check the remaining failed items.

In [ ]:
missing_geo = customer_orders[customer_orders["Lat"].isna()]
print("Still missing geocoding:", len(missing_geo))
display(missing_geo[["Customer ID", "Customer Address", "Clean_Address"]])

Still missing geocoding: 3


,Customer ID,Customer Address,Clean_Address
0,CUST1667,"Garden Plan, San Leandro, CA 94578","Garden Plan, San Leandro, CA 94578"
4,CUST3446,"1515 165th Ave, San Leandro, CA 94578","1515 165th Ave, San Leandro, CA 94578"
6,CUST4340,"28 Buckelew Street, Sausalito, CA 94965","28 Buckelew Street, Sausalito, CA 94965"


#The number of unread coordinates has now dropped to three. I manually located them to obtain the precise coordinates, then entered them directly.
#Customer CUST1667's address ("Garden Plan, San Leandro, CA 94578") could not be geocoded due to its non-standard format. Consequently, we employed the postal code centroid method to position this customer at the geographic center of postal code 94578.

In [ ]:
# Manually enter precise coordinates
customer_orders.loc[customer_orders["Customer ID"] == "CUST3446", "Lat"] = 37.69557683447166
customer_orders.loc[customer_orders["Customer ID"] == "CUST3446", "Lon"] = -122.10998394453172

customer_orders.loc[customer_orders["Customer ID"] == "CUST4340", "Lat"] = 37.87410449555437
customer_orders.loc[customer_orders["Customer ID"] == "CUST4340", "Lon"] = -122.51642513103535

customer_orders.loc[customer_orders["Customer ID"] == "CUST1667", "Lat"] = 37.6990
customer_orders.loc[customer_orders["Customer ID"] == "CUST1667", "Lon"] = -122.1560

#Confirm that `customer_orders` is currently available.

In [ ]:
print(customer_orders.columns.tolist())
display(customer_orders.head())

['Customer ID', 'Total_Volume', 'Total_Weight', 'Total_Load_Time', 'Total_Unload_Time', 'Total_Items', 'Total_Service_Time', 'Customer Address', 'Customer Address_y', 'Location', 'Lat', 'Lon', 'Clean_Address', 'Location_retry']


,Customer ID,Total_Volume,Total_Weight,Total_Load_Time,Total_Unload_Time,Total_Items,Total_Service_Time,Customer Address,Customer Address_y,Location,Lat,Lon,Clean_Address,Location_retry
0,CUST1667,1024.68,5053,56.0,100.0,38,156.0,"Garden Plan, San Leandro, CA 94578","Garden Plan, San Leandro, CA 94578",None,37.699000,-122.156000,"Garden Plan, San Leandro, CA 94578",None
1,CUST2208,189.05,464,9.0,13.0,3,22.0,"912 Sienna Loop, San Leandro, CA 94579","912 Sienna Loop, San Leandro, CA 94579","(Sienna Loop, San Leandro, Alameda County, Cal...",37.692196,-122.146782,"912 Sienna Loop, San Leandro, CA 94579",NaN
2,CUST2289,1274.12,4211,59.5,96.0,33,155.5,"4 Captain Dr Unit E211, Emeryville, CA 94608","4 Captain Dr Unit E211, Emeryville, CA 94608",None,37.838388,-122.303434,"4 Captain Dr , Emeryville, CA 94608","(4, Captain Drive, Emeryville, Alameda County,..."
3,CUST3051,1860.07,5404,81.5,147.0,43,228.5,"1510 165th Ave, San Leandro, CA 94578","1510 165th Ave, San Leandro, CA 94578","(1510, 165th Avenue, San Leandro, Alameda Coun...",37.694998,-122.109606,"1510 165th Ave, San Leandro, CA 94578",NaN
4,CUST3446,2133.54,5904,83.5,155.5,35,239.0,"1515 165th Ave, San Leandro, CA 94578","1515 165th Ave, San Leandro, CA 94578",None,37.695577,-122.109984,"1515 165th Ave, San Leandro, CA 94578",None


#Calculate the approximate distance from the warehouse to the customer.
##Define warehouse coordinates and distance function:

In [ ]:
from math import sqrt
import numpy as np
import pandas as pd

warehouse_lat = 37.759
warehouse_lon = -122.388

def approx_miles(lat1, lon1, lat2, lon2):
    return sqrt((lat1 - lat2)**2 + (lon1 - lon2)**2) * 69

#Calculate the distance of each customer from the warehouse:

In [ ]:
customer_orders["Miles_From_Warehouse"] = customer_orders.apply(
    lambda row: approx_miles(
        warehouse_lat,
        warehouse_lon,
        row["Lat"],
        row["Lon"]
    ),
    axis=1
)

display(customer_orders[[
    "Customer ID", "Customer Address", "Miles_From_Warehouse"
]].head(20))

,Customer ID,Customer Address,Miles_From_Warehouse
0,CUST1667,"Garden Plan, San Leandro, CA 94578",16.534681
1,CUST2208,"912 Sienna Loop, San Leandro, CA 94579",17.270519
2,CUST2289,"4 Captain Dr Unit E211, Emeryville, CA 94608",8.003364
3,CUST3051,"1510 165th Ave, San Leandro, CA 94578",19.710292
4,CUST3446,"1515 165th Ave, San Leandro, CA 94578",19.675943
5,CUST4225,"2125 166th Ave, San Leandro, CA 94578",20.135882
6,CUST4340,"28 Buckelew Street, Sausalito, CA 94965",11.899661
7,CUST4472,"1530 165th Ave, San Leandro, CA 94578",19.731367
8,CUST5271,"8234 Jackson St, San Francisco, CA 94109 (Nob...",3.279987
9,CUST5533,"1415 La Loma Ave, Berkeley, CA 94708",12.410272


#Check if any single client exceeds the capacity limit.

In [ ]:
MAX_VOLUME = 1700
MAX_WEIGHT = 7000
MAX_MINUTES = 540
AVG_SPEED_MPH = 30

customer_orders["Requires_Split_Delivery"] = np.where(
    (customer_orders["Total_Volume"] > MAX_VOLUME) |
    (customer_orders["Total_Weight"] > MAX_WEIGHT),
    "Yes",
    "No"
)

oversized_customers = customer_orders[
    customer_orders["Requires_Split_Delivery"] == "Yes"
]

print("Oversized customers:", len(oversized_customers))
display(oversized_customers[[
    "Customer ID", "Total_Volume", "Total_Weight", "Customer Address"
]])

Oversized customers: 5


,Customer ID,Total_Volume,Total_Weight,Customer Address
3,CUST3051,1860.07,5404,"1510 165th Ave, San Leandro, CA 94578"
4,CUST3446,2133.54,5904,"1515 165th Ave, San Leandro, CA 94578"
11,CUST5981,2414.60,6439,"208 Spencer Avenue, Sausalito, CA 94965"
12,CUST6434,1898.57,5742,"1 Captain Dr Unit D255, Emeryville, CA 94608"
15,CUST8286,1726.42,6721,"45 Hill Rd, Berkeley, CA 94708"


#Generate the "normal customers" subset (those whose orders can be fulfilled by a single vehicle).

In [ ]:
regular_customers = customer_orders[
    customer_orders["Requires_Split_Delivery"] == "No"
].copy()

regular_customers = regular_customers.sort_values("Miles_From_Warehouse").reset_index(drop=True)

print("Regular customers:", len(regular_customers))
display(regular_customers[[
    "Customer ID", "Miles_From_Warehouse", "Total_Volume",
    "Total_Weight", "Total_Service_Time"
]])

Regular customers: 14


,Customer ID,Miles_From_Warehouse,Total_Volume,Total_Weight,Total_Service_Time
0,CUST5271,3.279987,684.20,2056,58.5
1,CUST7988,3.279987,96.31,474,26.0
2,CUST8391,4.538845,1009.41,2925,103.0
3,CUST2289,8.003364,1274.12,4211,155.5
4,CUST7228,9.715911,524.07,1555,115.0
5,CUST5710,11.284318,974.29,4935,231.5
6,CUST4340,11.899661,980.88,5112,175.5
7,CUST8762,12.203277,413.01,618,33.0
8,CUST5533,12.410272,876.78,3685,127.5
9,CUST8551,12.806799,17.50,211,16.0


In [ ]:
unassigned = regular_customers.copy()

# To make it easier to fill up the day's schedule, prioritize tasks based on a combined assessment of "service duration + volume + weight," giving precedence to the largest items.
unassigned["Sort_Score"] = (
    unassigned["Total_Service_Time"] * 1.0 +
    unassigned["Total_Volume"] * 0.1 +
    unassigned["Total_Weight"] * 0.01
)

unassigned = unassigned.sort_values("Sort_Score", ascending=False).reset_index(drop=True)

display(unassigned[[
    "Customer ID", "Total_Volume", "Total_Weight",
    "Total_Service_Time", "Miles_From_Warehouse", "Sort_Score"
]])

,Customer ID,Total_Volume,Total_Weight,Total_Service_Time,Miles_From_Warehouse,Sort_Score
0,CUST5710,974.29,4935,231.5,11.284318,378.279
1,CUST2289,1274.12,4211,155.5,8.003364,325.022
2,CUST4340,980.88,5112,175.5,11.899661,324.708
3,CUST1667,1024.68,5053,156.0,16.534681,308.998
4,CUST5533,876.78,3685,127.5,12.410272,252.028
5,CUST8391,1009.41,2925,103.0,4.538845,233.191
6,CUST7228,524.07,1555,115.0,9.715911,182.957
7,CUST5271,684.20,2056,58.5,3.279987,147.480
8,CUST4225,343.33,1815,78.5,20.135882,130.983
9,CUST8762,413.01,618,33.0,12.203277,80.481


#Installing OR-Tools

In [ ]:
!pip install ortools

#Import OR-Tools

In [ ]:
from ortools.constraint_solver import pywrapcp, routing_enums_pb2

#Define OR-Tools Routing Functions

In [ ]:
def build_distance_matrix(day_df):
    locations = [(warehouse_lat, warehouse_lon)] + list(zip(day_df["Lat"], day_df["Lon"]))

    matrix = []
    for i in range(len(locations)):
        row = []
        for j in range(len(locations)):
            dist = approx_miles(
                locations[i][0], locations[i][1],
                locations[j][0], locations[j][1]
            )
            row.append(int(dist * 1000))  # 转成整数
        matrix.append(row)
    return matrix

#Define the solution function.

In [ ]:
def solve_daily_route(day_df):
    distance_matrix = build_distance_matrix(day_df)

    manager = pywrapcp.RoutingIndexManager(len(distance_matrix), 1, 0)
    routing = pywrapcp.RoutingModel(manager)

    def distance_callback(from_index, to_index):
        from_node = manager.IndexToNode(from_index)
        to_node = manager.IndexToNode(to_index)
        return distance_matrix[from_node][to_node]

    transit_callback_index = routing.RegisterTransitCallback(distance_callback)
    routing.SetArcCostEvaluatorOfAllVehicles(transit_callback_index)

    search_parameters = pywrapcp.DefaultRoutingSearchParameters()
    search_parameters.first_solution_strategy = routing_enums_pb2.FirstSolutionStrategy.PATH_CHEAPEST_ARC

    solution = routing.SolveWithParameters(search_parameters)

    if not solution:
        return None

    route_nodes = []
    index = routing.Start(0)

    while not routing.IsEnd(index):
        node = manager.IndexToNode(index)
        route_nodes.append(node)
        index = solution.Value(routing.NextVar(index))

    route_nodes.append(manager.IndexToNode(index))
    return route_nodes

#The First Edition of "Divisional Logic"

In [ ]:
def calculate_day_metrics(day_customer_ids, customer_df, warehouse_lat, warehouse_lon, avg_speed_mph=30):
    """
Given a list of customer IDs for a specific day:
1. Use OR-Tools to determine the optimal sequence.
2. Calculate the actual total mileage, total travel time, total service time, and total working hours.
    """
    if len(day_customer_ids) == 0:
        return {
            "ordered_customers": [],
            "total_drive_miles": 0,
            "total_drive_minutes": 0,
            "total_service_minutes": 0,
            "total_work_minutes": 23
        }

    day_df = customer_df[customer_df["Customer ID"].isin(day_customer_ids)].copy().reset_index(drop=True)

    route_nodes = solve_daily_route(day_df)
    if route_nodes is None:
        return None

    ordered_customers = [
        day_df.iloc[node - 1]["Customer ID"]
        for node in route_nodes
        if node != 0
    ]

    ordered_df = customer_df[customer_df["Customer ID"].isin(ordered_customers)].copy()
    ordered_df["route_order"] = ordered_df["Customer ID"].map(
        {cust_id: i for i, cust_id in enumerate(ordered_customers)}
    )
    ordered_df = ordered_df.sort_values("route_order").reset_index(drop=True)

    total_drive_miles = 0
    total_drive_minutes = 0
    total_service_minutes = 0

    prev_lat, prev_lon = warehouse_lat, warehouse_lon

    for _, row in ordered_df.iterrows():
        leg_miles = approx_miles(prev_lat, prev_lon, row["Lat"], row["Lon"])
        leg_minutes = leg_miles / avg_speed_mph * 60

        total_drive_miles += leg_miles
        total_drive_minutes += leg_minutes

        # Service Time per Stop = Total Loading/Unloading Time + Fixed Stop Time
        stop_fixed = 14
        service_minutes = row["Total_Service_Time"] + stop_fixed
        total_service_minutes += service_minutes

        prev_lat, prev_lon = row["Lat"], row["Lon"]

    # Final stop: back to the warehouse.
    return_miles = approx_miles(prev_lat, prev_lon, warehouse_lat, warehouse_lon)
    return_minutes = return_miles / avg_speed_mph * 60

    total_drive_miles += return_miles
    total_drive_minutes += return_minutes

    total_work_minutes = 23 + total_drive_minutes + total_service_minutes

    return {
        "ordered_customers": ordered_customers,
        "total_drive_miles": total_drive_miles,
        "total_drive_minutes": total_drive_minutes,
        "total_service_minutes": total_service_minutes,
        "total_work_minutes": total_work_minutes
    }

In [ ]:
MAX_VOLUME = 1700
MAX_WEIGHT = 7000
MAX_MINUTES = 540
AVG_SPEED_MPH = 30

days = []
remaining_ids = unassigned["Customer ID"].tolist()

while remaining_ids:
    # First, select a seed customer (the largest one remaining).
    seed_id = remaining_ids[0]
    current_day = [seed_id]
    remaining_ids.remove(seed_id)

    improved = True

    while improved and remaining_ids:
        improved = False
        best_candidate = None
        best_metrics = None
        best_score = -1

        for cust_id in remaining_ids:
            candidate_day = current_day + [cust_id]

            candidate_df = regular_customers[
                regular_customers["Customer ID"].isin(candidate_day)
            ].copy()

            total_volume = candidate_df["Total_Volume"].sum()
            total_weight = candidate_df["Total_Weight"].sum()

            # First, check the capacity.
            if total_volume > MAX_VOLUME or total_weight > MAX_WEIGHT:
                continue

            # Re-verify actual working hours (including travel time between clients + return to depot).
            metrics = calculate_day_metrics(
                candidate_day,
                regular_customers,
                warehouse_lat,
                warehouse_lon,
                AVG_SPEED_MPH
            )

            if metrics is None:
                continue

            if metrics["total_work_minutes"] <= MAX_MINUTES:
                # Scoring: Prioritize making the day "fuller"
                # Think of it this way: the closer to 540, the better.
                score = metrics["total_work_minutes"]

                if score > best_score:
                    best_score = score
                    best_candidate = cust_id
                    best_metrics = metrics

        if best_candidate is not None:
            current_day.append(best_candidate)
            remaining_ids.remove(best_candidate)
            improved = True

    days.append(current_day)

for i, d in enumerate(days, 1):
    print(f"Day {i}: {d}")

Day 1: ['CUST5710', 'CUST4225', 'CUST8551']
Day 2: ['CUST2289', 'CUST2208', 'CUST7988', 'CUST4472']
Day 3: ['CUST4340', 'CUST7228']
Day 4: ['CUST1667', 'CUST8762']
Day 5: ['CUST5533', 'CUST5271']
Day 6: ['CUST8391']


In [ ]:
def pack_trips_into_workdays(trips, customer_df, warehouse_lat, warehouse_lon, max_minutes=540, avg_speed_mph=30):
    """
    Re-pack existing trips into true workdays.
    Logic:
    - Treat each element in `trips` as one complete truck trip:
        Warehouse -> customers -> Warehouse
    - Multiple trips can be assigned to the same workday
      as long as cumulative work time <= max_minutes.

    Inputs
    ------
    trips : list[list[str]]
        Each inner list is a trip made of customer IDs.
    customer_df : pd.DataFrame
        Must contain Customer ID, Total_Volume, Total_Weight, Total_Service_Time, Lat, Lon.
    warehouse_lat, warehouse_lon : float
        Depot coordinates.
    max_minutes : int
        Max total work minutes per day, default 540.
    avg_speed_mph : float
        Travel speed assumption.

    Returns
    -------
    workdays : list[list[list[str]]]
        Structure:
        [
          [trip1, trip2],   # Workday 1 has 2 trips
          [trip3],          # Workday 2 has 1 trip
          ...
        ]
    """

    def trip_metrics(customer_ids):
        """
        Compute full trip metrics:
        Warehouse -> customers -> Warehouse
        using OR-Tools route inside the trip.
        """
        if len(customer_ids) == 0:
            return {
                "ordered_customers": [],
                "total_drive_miles": 0,
                "total_drive_minutes": 0,
                "total_service_minutes": 0,
                "total_work_minutes": 0,
                "total_volume": 0,
                "total_weight": 0
            }

        trip_df = customer_df[customer_df["Customer ID"].isin(customer_ids)].copy()

        total_volume = trip_df["Total_Volume"].sum()
        total_weight = trip_df["Total_Weight"].sum()

        # Single trip itself still cannot violate truck capacity
        if total_volume > MAX_VOLUME or total_weight > MAX_WEIGHT:
            return None

        # Use existing OR-Tools routine to optimize the stop sequence
        day_df = trip_df.reset_index(drop=True)
        route_nodes = solve_daily_route(day_df)

        if route_nodes is None:
            return None

        ordered_customers = [
            day_df.iloc[node - 1]["Customer ID"]
            for node in route_nodes
            if node != 0
        ]

        ordered_df = customer_df[customer_df["Customer ID"].isin(ordered_customers)].copy()
        ordered_df["route_order"] = ordered_df["Customer ID"].map(
            {cust_id: i for i, cust_id in enumerate(ordered_customers)}
        )
        ordered_df = ordered_df.sort_values("route_order").reset_index(drop=True)

        total_drive_miles = 0
        total_drive_minutes = 0
        total_service_minutes = 0

        prev_lat, prev_lon = warehouse_lat, warehouse_lon

        for _, row in ordered_df.iterrows():
            leg_miles = approx_miles(prev_lat, prev_lon, row["Lat"], row["Lon"])
            leg_minutes = leg_miles / avg_speed_mph * 60

            total_drive_miles += leg_miles
            total_drive_minutes += leg_minutes

            stop_fixed = 14
            service_minutes = row["Total_Service_Time"] + stop_fixed
            total_service_minutes += service_minutes

            prev_lat, prev_lon = row["Lat"], row["Lon"]

        # return to warehouse
        return_miles = approx_miles(prev_lat, prev_lon, warehouse_lat, warehouse_lon)
        return_minutes = return_miles / avg_speed_mph * 60

        total_drive_miles += return_miles
        total_drive_minutes += return_minutes

        # include 23 min startup for each trip
        total_work_minutes = 23 + total_drive_minutes + total_service_minutes

        return {
            "ordered_customers": ordered_customers,
            "total_drive_miles": total_drive_miles,
            "total_drive_minutes": total_drive_minutes,
            "total_service_minutes": total_service_minutes,
            "total_work_minutes": total_work_minutes,
            "total_volume": total_volume,
            "total_weight": total_weight
        }

    # First compute metrics for every existing trip
    trip_pool = []
    for idx, trip in enumerate(trips, start=1):
        metrics = trip_metrics(trip)
        if metrics is None:
            print(f"Trip {idx} infeasible and skipped: {trip}")
            continue

        trip_pool.append({
            "trip_id": idx,
            "customers": metrics["ordered_customers"],
            "trip_work_minutes": metrics["total_work_minutes"],
            "trip_drive_miles": metrics["total_drive_miles"],
            "trip_drive_minutes": metrics["total_drive_minutes"],
            "trip_service_minutes": metrics["total_service_minutes"],
            "trip_volume": metrics["total_volume"],
            "trip_weight": metrics["total_weight"]
        })

    # Sort trips by descending trip duration so large trips are packed first
    trip_pool = sorted(trip_pool, key=lambda x: x["trip_work_minutes"], reverse=True)

    workdays = []

    for trip in trip_pool:
        best_day_idx = None
        best_remaining_time = None

        for i, wd in enumerate(workdays):
            current_minutes = sum(t["trip_work_minutes"] for t in wd)

            if current_minutes + trip["trip_work_minutes"] <= max_minutes:
                remaining_time = max_minutes - (current_minutes + trip["trip_work_minutes"])

                # Best fit: choose the day that leaves the least remaining time
                if best_remaining_time is None or remaining_time < best_remaining_time:
                    best_remaining_time = remaining_time
                    best_day_idx = i

        if best_day_idx is None:
            # Start a new workday
            workdays.append([trip])
        else:
            # Append to best existing day
            workdays[best_day_idx].append(trip)

    return workdays

In [ ]:
trips = days

workdays = pack_trips_into_workdays(
    trips,
    regular_customers,
    warehouse_lat,
    warehouse_lon,
    MAX_MINUTES,
    AVG_SPEED_MPH
)

for wd_idx, wd in enumerate(workdays, start=1):
    print(f"\nWorkday {wd_idx}")
    total_minutes = sum(t["trip_work_minutes"] for t in wd)
    print(f"  Total work minutes: {total_minutes:.2f}")
    for t_idx, trip in enumerate(wd, start=1):
        print(f"  Trip {t_idx}: {trip['customers']} | {trip['trip_work_minutes']:.2f} min")


Workday 1
  Total work minutes: 521.91
  Trip 1: ['CUST5710', 'CUST8551', 'CUST4225'] | 521.91 min

Workday 2
  Total work minutes: 424.34
  Trip 1: ['CUST7228', 'CUST4340'] | 424.34 min

Workday 3
  Total work minutes: 387.26
  Trip 1: ['CUST7988', 'CUST2289', 'CUST4472', 'CUST2208'] | 387.26 min

Workday 4
  Total work minutes: 480.20
  Trip 1: ['CUST8762', 'CUST1667'] | 322.05 min
  Trip 2: ['CUST8391'] | 158.16 min

Workday 5
  Total work minutes: 293.86
  Trip 1: ['CUST5271', 'CUST5533'] | 293.86 min


In [ ]:
detailed_rows = []

for workday_num, workday in enumerate(workdays, start=1):
    current_day_time = 8 * 60  # 08:00 开始一天

    for trip_num, trip in enumerate(workday, start=1):
        trip_customer_ids = trip["customers"]

        trip_df = regular_customers[
            regular_customers["Customer ID"].isin(trip_customer_ids)
        ].copy()

        trip_df["route_order"] = trip_df["Customer ID"].map(
            {cust_id: i for i, cust_id in enumerate(trip_customer_ids)}
        )
        trip_df = trip_df.sort_values("route_order").reset_index(drop=True)

        # Preparation/Loading at the start of each trip: 23 minutes.
        current_day_time += 23

        prev_lat, prev_lon = warehouse_lat, warehouse_lon

        for stop_order, (_, row) in enumerate(trip_df.iterrows(), start=1):
            drive_miles = approx_miles(prev_lat, prev_lon, row["Lat"], row["Lon"])
            drive_minutes = drive_miles / AVG_SPEED_MPH * 60

            arrival_time = current_day_time + drive_minutes
            waiting_time = 0

            stop_fixed = 14
            service_minutes = row["Total_Service_Time"] + stop_fixed
            finish_time = arrival_time + service_minutes

            detailed_rows.append({
                "Workday": workday_num,
                "Trip_Number": trip_num,
                "Stop_Order": stop_order,
                "Customer ID": row["Customer ID"],
                "Customer Address": row["Customer Address"],
                "Arrival_Minutes": arrival_time,
                "Finish_Minutes": finish_time,
                "Arrival_Time": minutes_to_time(arrival_time),
                "Finish_Time": minutes_to_time(finish_time),
                "Drive_Miles_From_Previous": drive_miles,
                "Drive_Minutes_From_Previous": drive_minutes,
                "Waiting_Minutes": waiting_time,
                "Service_Minutes": service_minutes,
                "Total_Volume": row["Total_Volume"],
                "Total_Weight": row["Total_Weight"],
                "Lat": row["Lat"],
                "Lon": row["Lon"]
            })

            current_day_time = finish_time
            prev_lat, prev_lon = row["Lat"], row["Lon"]

        # Return to the warehouse after each trip.
        return_miles = approx_miles(prev_lat, prev_lon, warehouse_lat, warehouse_lon)
        return_minutes = return_miles / AVG_SPEED_MPH * 60
        return_arrival = current_day_time + return_minutes

        detailed_rows.append({
            "Workday": workday_num,
            "Trip_Number": trip_num,
            "Stop_Order": "Return",
            "Customer ID": "WAREHOUSE",
            "Customer Address": "590 20th St, San Francisco, CA 94107",
            "Arrival_Minutes": return_arrival,
            "Finish_Minutes": return_arrival,
            "Arrival_Time": minutes_to_time(return_arrival),
            "Finish_Time": minutes_to_time(return_arrival),
            "Drive_Miles_From_Previous": return_miles,
            "Drive_Minutes_From_Previous": return_minutes,
            "Waiting_Minutes": 0,
            "Service_Minutes": 0,
            "Total_Volume": 0,
            "Total_Weight": 0,
            "Lat": warehouse_lat,
            "Lon": warehouse_lon
        })

        current_day_time = return_arrival

schedule_detailed = pd.DataFrame(detailed_rows)
schedule_detailed["Stop_Sort"] = schedule_detailed.groupby(["Workday", "Trip_Number"]).cumcount() + 1

display(schedule_detailed.head(50))

,Workday,Trip_Number,Stop_Order,Customer ID,Customer Address,Arrival_Minutes,Finish_Minutes,Arrival_Time,Finish_Time,Drive_Miles_From_Previous,Drive_Minutes_From_Previous,Waiting_Minutes,Service_Minutes,Total_Volume,Total_Weight,Lat,Lon,Stop_Sort
0,1,1,1,CUST5710,"410 Donahue Street, Sausalito, CA 94965",525.568636,771.068636,08:46,12:51,11.284318,22.568636,0,245.5,974.29,4935,37.872444,-122.505797,1
1,1,1,2,CUST8551,"1350 Queens Rd, Berkeley, CA 94708",806.102954,836.102954,13:26,13:56,17.517159,35.034318,0,30.0,17.50,211,37.885593,-122.252266,2
2,1,1,3,CUST4225,"2125 166th Ave, San Leandro, CA 94578",869.139968,961.639968,14:29,16:02,16.518507,33.037014,0,92.5,343.33,1815,37.698877,-122.102436,3
3,1,1,Return,WAREHOUSE,"590 20th St, San Francisco, CA 94107",1001.911732,1001.911732,16:42,16:42,20.135882,40.271764,0,0.0,0.00,0,37.759000,-122.388000,4
4,2,1,1,CUST7228,"2388 Stanford Street, Alameda, CA 94501",522.431822,651.431822,08:42,10:51,9.715911,19.431822,0,129.0,524.07,1555,37.772877,-122.247875,1
5,2,1,2,CUST4340,"28 Buckelew Street, Sausalito, CA 94965",691.037098,880.537098,11:31,14:41,19.802638,39.605276,0,189.5,980.88,5112,37.874104,-122.516425,2
6,2,1,Return,WAREHOUSE,"590 20th St, San Francisco, CA 94107",904.336421,904.336421,15:04,15:04,11.899661,23.799323,0,0.0,0.00,0,37.759000,-122.388000,3
7,3,1,1,CUST7988,"4203 Jackson St, San Francisco, CA 94109 (Nob...",509.559974,549.559974,08:30,09:10,3.279987,6.559974,0,40.0,96.31,474,37.794357,-122.419773,1
8,3,1,2,CUST2289,"4 Captain Dr Unit E211, Emeryville, CA 94608",566.726131,736.226131,09:27,12:16,8.583079,17.166157,0,169.5,1274.12,4211,37.838388,-122.303434,2
9,3,1,3,CUST4472,"1530 165th Ave, San Leandro, CA 94578",769.530146,791.530146,12:50,13:12,16.652007,33.304015,0,22.0,16.33,325,37.695048,-122.109281,3


In [ ]:
LABOR_PER_HOUR = 104
TRUCK_COST_PER_MILE = 1.205
CO2_PER_MILE = 1.69
CO2_IDLE_PER_HOUR = 3.97

daily_summary = schedule_detailed.groupby("Workday").agg(
    Total_Drive_Miles=("Drive_Miles_From_Previous", "sum"),
    Total_Drive_Minutes=("Drive_Minutes_From_Previous", "sum"),
    Total_Waiting_Minutes=("Waiting_Minutes", "sum"),
    Total_Service_Minutes=("Service_Minutes", "sum"),
    Stops=("Customer ID", lambda x: (x != "WAREHOUSE").sum()),
    Trips=("Trip_Number", "nunique")
).reset_index()

# Each trip includes a 23-minute preparation period.
trip_counts = schedule_detailed[schedule_detailed["Customer ID"] == "WAREHOUSE"].groupby("Workday").size().reset_index(name="Trip_Count")
daily_summary = daily_summary.merge(trip_counts, on="Workday", how="left")

daily_summary["Total_Work_Minutes"] = (
    daily_summary["Trip_Count"] * 23
    + daily_summary["Total_Drive_Minutes"]
    + daily_summary["Total_Service_Minutes"]
)

daily_summary["Labor_Cost"] = (daily_summary["Total_Work_Minutes"] / 60) * LABOR_PER_HOUR
daily_summary["Truck_Cost"] = daily_summary["Total_Drive_Miles"] * TRUCK_COST_PER_MILE
daily_summary["Total_Cost"] = daily_summary["Labor_Cost"] + daily_summary["Truck_Cost"]

daily_summary["CO2_Drive"] = daily_summary["Total_Drive_Miles"] * CO2_PER_MILE
daily_summary["CO2_Idle"] = (daily_summary["Total_Service_Minutes"] / 60) * CO2_IDLE_PER_HOUR
daily_summary["Total_CO2"] = daily_summary["CO2_Drive"] + daily_summary["CO2_Idle"]

daily_summary["Utilization_%"] = daily_summary["Total_Work_Minutes"] / 540

display(daily_summary)

,Workday,Total_Drive_Miles,Total_Drive_Minutes,Total_Waiting_Minutes,Total_Service_Minutes,Stops,Trips,Trip_Count,Total_Work_Minutes,Labor_Cost,Truck_Cost,Total_Cost,CO2_Drive,CO2_Idle,Total_CO2,Utilization_%
0,1,65.455866,130.911732,0,368.0,3,1,1,521.911732,904.647002,78.874319,983.521321,110.620414,24.349333,134.969747,0.966503
1,2,41.418211,82.836421,0,318.5,2,1,1,424.336421,735.516464,49.908944,785.425408,69.996776,21.074083,91.070859,0.785808
2,3,48.380654,96.761308,0,267.5,4,1,1,387.261308,671.252933,58.298688,729.551621,81.763305,17.699583,99.462888,0.717151
3,4,50.101352,100.202703,0,334.0,3,2,2,480.202703,832.351352,60.372129,892.723481,84.671284,22.099667,106.770951,0.889264
4,5,28.427865,56.855729,0,214.0,2,1,1,293.855729,509.349930,34.255577,543.605507,48.043091,14.159667,62.202758,0.544177


#Clean Up Fields

In [ ]:
daily_summary = daily_summary.drop(columns=["Trips"])

#Utilization Level

In [ ]:
def utilization_label(x):
    if x >= 0.9:
        return "Excellent"
    elif x >= 0.75:
        return "Good"
    elif x >= 0.5:
        return "Moderate"
    else:
        return "Low"

daily_summary["Utilization_Level"] = daily_summary["Utilization_%"].apply(utilization_label)

Export workday_trip_summary_final.csv

In [ ]:
workday_trip_summary = schedule_detailed.groupby(
    ["Workday", "Trip_Number"]
).agg(
    Trip_Drive_Miles=("Drive_Miles_From_Previous", "sum"),
    Trip_Drive_Minutes=("Drive_Minutes_From_Previous", "sum"),
    Trip_Service_Minutes=("Service_Minutes", "sum"),
    Stops=("Customer ID", lambda x: (x != "WAREHOUSE").sum())
).reset_index()

# Export
workday_trip_summary.to_csv("workday_trip_summary_final.csv", index=False)

# Take a look at the results.
display(workday_trip_summary.head(20))

,Workday,Trip_Number,Trip_Drive_Miles,Trip_Drive_Minutes,Trip_Service_Minutes,Stops
0,1,1,65.455866,130.911732,368.0,3
1,2,1,41.418211,82.836421,318.5,2
2,3,1,48.380654,96.761308,267.5,4
3,4,1,41.023662,82.047323,217.0,2
4,4,2,9.077690,18.155380,117.0,1
5,5,1,28.427865,56.855729,214.0,2


Export oversized_customers_final.csv

In [ ]:
oversized_customers.to_csv("oversized_customers_final.csv", index=False)

display(oversized_customers.head())

,Customer ID,Total_Volume,Total_Weight,Total_Load_Time,Total_Unload_Time,Total_Items,Total_Service_Time,Customer Address,Customer Address_y,Location,Lat,Lon,Clean_Address,Location_retry,Miles_From_Warehouse,Requires_Split_Delivery
3,CUST3051,1860.07,5404,81.5,147.0,43,228.5,"1510 165th Ave, San Leandro, CA 94578","1510 165th Ave, San Leandro, CA 94578","(1510, 165th Avenue, San Leandro, Alameda Coun...",37.694998,-122.109606,"1510 165th Ave, San Leandro, CA 94578",NaN,19.710292,Yes
4,CUST3446,2133.54,5904,83.5,155.5,35,239.0,"1515 165th Ave, San Leandro, CA 94578","1515 165th Ave, San Leandro, CA 94578",None,37.695577,-122.109984,"1515 165th Ave, San Leandro, CA 94578",None,19.675943,Yes
11,CUST5981,2414.60,6439,80.0,129.5,35,209.5,"208 Spencer Avenue, Sausalito, CA 94965","208 Spencer Avenue, Sausalito, CA 94965","(208, Spencer Avenue, Sausalito, Marin County,...",37.853389,-122.488299,"208 Spencer Avenue, Sausalito, CA 94965",NaN,9.503256,Yes
12,CUST6434,1898.57,5742,65.0,103.0,28,168.0,"1 Captain Dr Unit D255, Emeryville, CA 94608","1 Captain Dr Unit D255, Emeryville, CA 94608",None,37.837281,-122.303107,"1 Captain Dr , Emeryville, CA 94608","(1, Captain Drive, Emeryville, Alameda County,...",7.967853,Yes
15,CUST8286,1726.42,6721,62.0,98.5,30,160.5,"45 Hill Rd, Berkeley, CA 94708","45 Hill Rd, Berkeley, CA 94708","(45, Hill Road, La Loma Park, Berkeley, Alamed...",37.889515,-122.251356,"45 Hill Rd, Berkeley, CA 94708",NaN,13.038241,Yes


#Export Final Table
####The files to be ultimately submitted to/imported into Power BI should be:

####customer_orders_final.csv
####schedule_detailed_final.csv
####daily_summary_final.csv
####workday_trip_summary_final.csv
####oversized_customers_final.csv

In [ ]:
schedule_detailed.to_csv("schedule_detailed_final.csv", index=False)
daily_summary.to_csv("daily_summary_final.csv", index=False)
customer_orders.to_csv("customer_orders_final.csv", index=False)
workday_trip_summary.to_csv("workday_trip_summary_final.csv", index=False)
oversized_customers.to_csv("oversized_customers_final.csv", index=False)

## Naive Baseline Comparison (Optimization Impact)

To quantify the value of the optimization, we compare the optimized plan against a **naive no-consolidation baseline**: one dedicated round trip per order (no batching, no route sequencing), costed with the **same** labor, fuel, and CO2 model and identical service times. Only routing efficiency differs.

In [ ]:
PREP_MIN = 23

# --- Optimized totals (from daily_summary) ---
opt_miles = daily_summary['Total_Drive_Miles'].sum()
opt_cost  = daily_summary['Total_Cost'].sum()
opt_co2   = daily_summary['Total_CO2'].sum()
opt_trips = int(daily_summary['Trip_Count'].sum())
service_min_total = daily_summary['Total_Service_Minutes'].sum()
idle_co2_total = (service_min_total / 60) * CO2_IDLE_PER_HOUR

# --- Naive baseline: one dedicated round trip per regular customer ---
naive_miles = regular_customers['Miles_From_Warehouse'].sum() * 2
naive_trips = len(regular_customers)
naive_drive_min = naive_miles * 2   # 30 mph (minutes = miles * 2)
naive_work_min  = naive_trips * PREP_MIN + naive_drive_min + service_min_total
naive_labor = naive_work_min / 60 * LABOR_PER_HOUR
naive_truck = naive_miles * TRUCK_COST_PER_MILE
naive_cost  = naive_labor + naive_truck
naive_co2   = naive_miles * CO2_PER_MILE + idle_co2_total

comparison = pd.DataFrame({
    'Scenario': ['Naive (no consolidation)', 'Optimized'],
    'Trips': [naive_trips, opt_trips],
    'Drive_Miles': [naive_miles, opt_miles],
    'Operating_Cost': [naive_cost, opt_cost],
    'Total_CO2_kg': [naive_co2, opt_co2],
})
for col in ['Drive_Miles', 'Operating_Cost', 'Total_CO2_kg']:
    saved = (comparison[col][0] - comparison[col][1]) / comparison[col][0] * 100
    print(f'{col:16} reduced {saved:5.1f}%  ({comparison[col][0]:.1f} -> {comparison[col][1]:.1f})')
print(f'Annualized cost saved (x52): ${(naive_cost-opt_cost)*52:,.0f}')
print(f'Annualized CO2 saved  (x52): {(naive_co2-opt_co2)*52:,.0f} kg')

comparison.to_csv('baseline_comparison_final.csv', index=False)
display(comparison)